### DISCUSSION TITLE
API interactions for dagRuns ignores global permissions

### DISCUSSION THREAD

**Discussion by benriou** [apache/airflow#33599] [General] [answered]
### Apache Airflow version

2.7.0

### What happened

After upgrading from Airflow 2.5 to 2.7.0, a service account is not allowed to perform API Calls to `https://<redacted>/api/v1/dags/<dag_name>/dagRuns` (403 error).

The role permission assigned to the service account was : `can create on DAG Runs`, `can edit on DAG Runs`, `can read on Website`, this was enough to start **any existing DAG** via the Airflow API.

Now we explicitly need to give per-dag permissions : `can edit on DAG:<dag_name>`, `can create on DAG:<dag_name>`.  Then, it's working again.

### What you think should happen instead

I did not notice any change between 2.5 to 2.7 related to the way Airflow handles role permissions. 
I suppose that "global permissions" (`can create on DAG Runs`) without explicitly defining a dag_name, should continue to operate like before. 

### How to reproduce

Create a user with a permission.
The permission should contain `can create on DAG Runs`, `can edit on DAG Runs`. 

Perform an API call as follows 
```
    export EXECUTION_DATE=$(date +"%Y-%m-%dT%H:%M:%S+0000")
    export UUID=$(cat /proc/sys/kernel/random/uuid)
    echo "Airflow execution date is: " $EXECUTION_DATE
    echo "dag_run_id_is : " $UUID
    curl -X 'POST' \
    'https://<AIRFLOW_HOSTNAME>/api/v1/dags/<DAG_NAME>/dagRuns' \
    -H 'accept: application/json' \
    -H 'Content-Type: application/json' \
    --fail-with-body \
    --user "CREDENTIALS_USER_NAME:CREDENTIALS_SECRET_KEY" \
    -d '{
    "conf": {},
    "dag_run_id": "'"$UUID"'"
    }'
~
```

You'll get a 403 error like : 
```
Airflow execution date is:  2023-08-21T15:10:35+0000
dag_run_id_is :
{
  "detail": null,
  "status": 403,
  "title": "Forbidden",
  "type": "https://airflow.apache.org/docs/apache-airflow/2.7.0/stable-rest-api-ref.html#section/Errors/PermissionDenied"
}
```

Edit the permissions assigned to the role, and add an explicit `can edit on DAG:<dag_name>` then the 403 is gone. 

### Operating System

Debian 11

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by SamWheating** [comment]
I have some time today, feel free to assign this to me and I'll see if I can replicate + fix.

**Comment by potiuk** [comment]
Thanks! Assigned :)

**Comment by SamWheating** [comment, marked as answer]
Looking into this, it looks like the access required to `POST /api/v1/dags/<dag_id>/dagRun` is `can edit on dag`, `can create on dagrun`:

https://github.com/apache/airflow/blob/319045492d2559bd856a43a1fa810adf59358d7d/airflow/api_connexion/endpoints/dag_run_endpoint.py#L306-L311

And in your example, you've got `can create on dagrun`, `can edit on dagrun`, which explains why you're getting 403'd (and  then subsequently succeeding after adding `can edit on dag:<dag_id>`)

I'd recommend adding `can edit on dag` to your service account. This worked for me when testing locally. Please confirm that this works and we can close the issue.

**Reply by benriou** [reply]
Hello, 

Thanks for looking into my issue. 
This is now working with `can edit on DAG` ; thank you ! 

We believe that this was not required before the upgrade. 
Did the permission model changed from 2.5 to 2.7 ? 

Thank you 
Ben

**Comment by potiuk** [comment]
Marked it as "pending response" - and it seems that the explanation from @SamWheating is super plausible and makes sense so I will convert into discussions. We can always convert it back to an issue if we have some new findings (Thanls @SamWheating for looking into it!)
